README.md
Boundary Apache HTTP Server Plugin
==================================

Collects metrics from a Apache HTTP server instance. See video [walkthrough](https://help.boundary.com/hc/articles/201991151).

### Prerequisites

|     OS    | Linux | Windows | SmartOS | OS X |
|:----------|:-----:|:-------:|:-------:|:----:|
| Supported |   v   |    v    |    v    |  v   |

#### Boundary Meter Versions V4.0 Or Greater

- To install new meter go to Settings->Installation or [see instructons](https://help.boundary.com/hc/en-us/sections/200634331-Installation).
- To upgrade the meter to the latest version - [see instructons](https://help.boundary.com/hc/en-us/articles/201573102-Upgrading-the-Boundary-Meter).

#### Boundary Meter Versions Less Than V4.0

|  Runtime | node.js | Python | Java |
|:---------|:-------:|:------:|:----:|
| Required |    +    |        |      |

- [How to install node.js?](https://help.boundary.com/hc/articles/202360701)
- Apache HTTP Server must be configured to run the `server-stats` module. 

### Plugin Setup

- The Boundary Apache HTTP Server plugin depends on the `server-stats` module for collecting metrics. The sections below provide the procedures to enable and configure the `server-stats` module.

#### Enable the `server-status` Module

1. Modify the Apache HTTP Server `httpd.conf` by adding the following:
     ```xml
     <Location /server-status>
		SetHandler server-status
	 </Location>
     ```

#### Secure the EndPoint with a User Name and Password
1. Create as password file to secure the endpoint. The example shown here is using the path `/etc/httpd/my_password_file`.
     ```
     $ sudo htpasswd -c /etc/httpd/my_password_file
     ```
2. Enable authentication by modifying the `<Location/>` added previously as shown here :
     ```xml
	<Location /server-status>
		SetHandler server-status
		AuthType basic
		AuthName "Apache status"
		AuthUserFile /etc/httpd/my_password_file
		Require valid-user
	</Location>
    ```
3. Restart Apache HTTP server reload the `httpd.conf` configuration.
4. Verify that statistics are being collected by visiting http://yourserver.com/server-status
5. To also capture the 'requests per second' metric, add `ExtendedStatus On` outside of your `<Location />` block:
     ```xml
	<Location /server-status>
		SetHandler server-status
		AuthType basic
		AuthName "Apache status"
		AuthUserFile /etc/httpd/my_password_file
		Require valid-user
	</Location>
	ExtendedStatus On
    ```

### Plugin Configuration Fields

|Field Name       |Description                                                                                  |
|:----------------|:--------------------------------------------------------------------------------------------|
|Server-Status URL|The URL endpoint of where the Apache HTPP server statistics are hosted.                      |
|Username         |If the URL is password protected, what username should the plugin use to authenticate        |
|Password         |If the URL is password protected, what password should the plugin use to authenticate        |
|Poll Interval    |How often (in milliseconds) to poll for metrics (default: 1000).      | 
|Source           |Name identifying the specific instance of Apache HTTP server which is displayed in dashboards|

### Metrics Collected
Tracks the following metrics for [apache](http://httpd.apache.org/)

|Metric Name              |Description                                      |
|:------------------------|:------------------------------------------------|
|Apache Requests          |The number of Apache Accesses                    |
|Apache Total Bytes       |bytes transferred                                |
|Apache Bytes per Request |average bytes per request                        |
|Apache CPU               |                                                 |
|Apache Busy Workers      |the number of busy workers                       |
|Apache Idle Workers      |the number of idle workers                       |
|Apache busy to idle ratio|The ratio of busy workers / (busy + idle workers)|

### Dashboards

Apache Server

### References

None


Changelog
* Changes from 1.0 to 1.1
  * Rewritten the plugin code to use the Boundary LUA Framework


init.lua
local framework = require('framework')
local url = require('url')
local Plugin = framework.Plugin
local WebRequestDataSource = framework.WebRequestDataSource
local Accumulator = framework.Accumulator

local gsplit = framework.string.gsplit
local split = framework.string.split
local auth = framework.util.auth

local params = framework.params

local options = url.parse(params.url)
options.path = options.path .. '?auto' -- server-status?auto for text/plain output
options.auth = auth(params.username, params.password) 
options.wait_for_end = true

local mapping = { 
  ['BusyWorkers'] = 'APACHE_BUSY_WORKERS',
  ['IdleWorkers'] = 'APACHE_IDLE_WORKERS',
  ['Total Accesses'] = 'APACHE_REQUESTS',
  ['CPULoad'] = 'APACHE_CPU',
  ['Total kBytes'] = 'APACHE_BYTES'
}
  
local data_source = WebRequestDataSource:new(options)
local acc = Accumulator:new()

local plugin = Plugin:new(params, data_source)
function plugin:onParseValues(data, _)

  -- Capture metrics
  local result = {}
  for v in gsplit(data, "\n") do
    local m = split(v, ":")
    local key, val = unpack(m)
    if (key and val) then
      local metric = mapping[key] 
      if metric then
        result[metric] = tonumber(val) or -1
      end
    end
  end

  -- CPU Load calculations
  local cpu_load = result['APACHE_CPU'] or 0
  if (cpu_load > 1) then
    cpu_load = cpu_load / 100
  end
  result['APACHE_CPU'] = cpu_load

  -- Requests calculation
  local requests = acc:accumulate('APACHE_REQUESTS', result['APACHE_REQUESTS']) 
  result['APACHE_REQUESTS'] = requests

  -- Total Bytes calculation
  -- Because of the interval cut off lines, on a really slow site you will get 0's
  -- the, use the previous value if that happens
  local last_total_bytes = acc:get('APACHE_BYTES')
  local total_bytes = acc:accumulate('APACHE_BYTES', result['APACHE_BYTES'] * 1024)
  if requests > 0 and total_bytes == 0 then
      total_bytes = last_total_bytes
  end
  result['APACHE_BYTES'] = total_bytes

  -- Total Bytes Per Request calculation
  local bytes_per_req = (requests > 0) and (total_bytes / requests) or 0
  result['APACHE_BYTES_PER_REQUEST'] = bytes_per_req

  -- Busy Ratio calculation
  local total_workers =  (result['APACHE_BUSY_WORKERS'] or 0) + (result['APACHE_IDLE_WORKERS'] or 0)
  local busy_ratio = total_workers and result['APACHE_BUSY_WORKERS'] / total_workers or 0
  result['APACHE_BUSY_RATIO'] = busy_ratio

  return result
end
plugin:run()


index.js
var _os = require('os');
var _param = require('./param.json');
var _request = require('request');
var _tools = require('graphdat-plugin-tools');

var _httpOptions; // username/password options for the URL
var _previous = {}; // remember the previous poll data so we can provide proper counts

// At a minimum, we need a way to contact apache
if (!_param.url)
{
    console.error('To get statistics from Apache, a statistics URL is required');
    process.exit(-1);
}

// if there is no http or https, guess its http
if (_param.url.substring(0,7) !== 'http://' && _param.url.substring(0,8) !== 'https://')
    _param.url = 'http://' + _param.url;

// tell apache to make the output 'machine readable'
if (_param.url.slice(-5) !== '?auto')
    _param.url += '?auto';

// This this is a URL and we have a name and password, then we need to add an auth header
if (_param.username)
    _httpOptions = { auth: { user: _param.username, pass: _param.password, sendImmediately: true } };

// If we do not have a source, we prefix everything with the servers hostname
_param.source = (_param.source || _os.hostname()).trim();

// If you do not have a poll intevarl, use 1000 as the default
_param.pollInterval = _param.pollInterval || 1000;

// get the natural difference between a and b
function diff(a, b)
{
    if (a == null || b == null)
        return 0;
    else
        return Math.max(a - b, 0);
}

function handleError(err)
{
    console.error(err);
    process.exit(-1);
}

function poll(cb)
{
    // call apache to get the stats page
    _request.get(_param.url, _httpOptions, function(err, resp, body)
    {
        if (err)
            return handleError(err);
        if (resp.statusCode !== 200)
           return handleError(new Error('Apache returned with an error - recheck the URL and credentials that you provided'));
        if (!body)
           return handleError(new Error('Apache statistics return empty'));

        var current = {};
        var lines = body.split('\n');
        lines.forEach(function(line)
        {
            // skip over empty lines
            if (!line)
                return;

           // parse the record
            var data = line.split(':');
            var key = data[0].trim();
            var value = data[1];

            if (value == null || value === '')
                return;

            var fValue = parseFloat(value,10);
            if (fValue === 0 || fValue)
                current[key] = fValue;
            else
                current[key] = value.trim();
        });

        var totalWorkers =  (current['BusyWorkers'] || 0) + (current['IdleWorkers'] || 0);
        var busyRatio = (totalWorkers) ? current['BusyWorkers'] / totalWorkers : 0;

        var requests = diff(current['Total Accesses'], _previous['Total Accesses']);

        // because of the interval cut off lines, on a really slow site you will get 0's
        // use the previous value if that happens
        current['totalBytes'] = diff(current['Total kBytes'], _previous['Total kBytes']) * 1024;
        if (requests > 0 && current['totalBytes'] ===  0)
            current['totalBytes'] = _previous['totalBytes'];

        var bytesPerReq = (requests) ? current['totalBytes']/requests : 0;
        var cpuLoad = current['CPULoad'] || 0;
        if (cpuLoad > 1)
            cpuLoad /= 100;

        console.log('APACHE_REQUESTS %d %s', requests, _param.source);
        console.log('APACHE_BYTES %d %s', current['totalBytes'], _param.source);
        console.log('APACHE_BYTES_PER_REQUEST %d %s', bytesPerReq, _param.source);
        console.log('APACHE_CPU %d %s', cpuLoad, _param.source);
        console.log('APACHE_BUSY_WORKERS %d %s', current['BusyWorkers'], _param.source);
        console.log('APACHE_IDLE_WORKERS %d %s', current['IdleWorkers'], _param.source);
        console.log('APACHE_BUSY_RATIO %d %s', busyRatio, _param.source);

        _previous = current;

        setTimeout(poll, _param.pollInterval);
    });
}

poll();
